Feature/amqp reconnect

Ashhm opened the pull request:
Reconnection mechanism implementation

Commit:
Implement reconnection mechanism
- Fix methods description
- Fix warnings
- Implement reconnection mechanism
- Add tests to cover reconnection implementation
- Add nyc for test coverage

Commit:
Implement reconnection mechanism
- Fix warnings

Commit:
Implement reconnection mechanism
- Fix eol

iredjee commented on src/index.js:
```diff
@@ -1,21 +1,122 @@
+       */
+      async _connect() {
+        let connectionTimeout;
+        const reconnect = () => {
```
Maybe extract this to methods?

Ashhm replied:
Done, check pls https://github.com/Ashhm/moleculer-amqp/pull/7/commits/319de97836594a87319cbc59729ad8861f072141

iredjee commented on src/index.js:
```diff
@@ -34,6 +135,7 @@ module.exports = function createService(url, options) {
         return this.channel
```
Maybe we need to check availability of rabbit and channel somehow, to ensure we are not in "reconnect" mode. WDYT?

iredjee replied:
And the same for sendToQueue

Ashhm replied:
I thought about that, probably we should have some mechanism to re-send all of these messages which are failed to send when reconnection is done

iredjee replied:
Maybe, it can be done separately then.

Ashhm replied:
Yea, need to brainstorm about how to deliver delayed response

Ashhm commented on src/index.js:
```diff
@@ -1,21 +1,122 @@
   return {
     name: 'amqp',
+    metadata: {
+      reconnectTimeout: 5000,
```
As I know, moleculer does not allow us to change this value in amqp transport (or disable reconnection at all), so I think we should not allow this also.

iredjee replied:
Let's follow their practice then )

Commit:
Implement reconnection mechanism
- Move reconnection to service methods
- Fix tests
- Connection setup should be init on service start hook

iredjee commented on src/index.js:
```diff
@@ -1,21 +1,119 @@
+       * @private
+       */
+      _reconnect() {
+        if (!this._isReconnecting) {
```
Looks like better approach !

iredjee commented on test/index.spec.js:
```diff
@@ -178,4 +185,227 @@ describe('AMQP queues', () => {
+      it('should set service channel', () => expect(service.channel).to.exist);
+    });
+
+    describe('on multiply errors', () => {
```
`on multiplE errors` is more correct :)

Ashhm replied:
Fixed

Commit:
Prepare to npm release
- Add README
- Add examples

Commit:
Prepare to npm release
- Add keywords

Commit:
Implement reconnection mechanism
- Fix typo
